src/herder/HerderImpl.cpp

```diff
@@ -845,17 +845,40 @@ HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger)
     auto const& lcl = mLedgerManager.getLastClosedLedgerHeader();
     auto proposedSet = mTransactionQueue.toTxSet(lcl);
 
-    auto upperBoundCloseTimeOffset =
-        getUpperBoundCloseTimeOffset(mApp, lcl.header.scpValue.closeTime);
+    // We pick as next close time the current time unless it's before the last
+    // close time. We don't know how much time it will take to reach consensus
+    // so this is the most appropriate value to use as closeTime.
+    uint64_t nextCloseTime =
+        VirtualClock::to_time_t(mApp.getClock().system_now());
+    if (nextCloseTime <= lcl.header.scpValue.closeTime)
+    {
+        nextCloseTime = lcl.header.scpValue.closeTime + 1;
+    }
 
-    auto removed = proposedSet->trimInvalid(mApp, upperBoundCloseTimeOffset);
+    // Protocols including the "closetime change" (CAP-0034) externalize
+    // the exact closeTime contained in the StellarValue with the best
+    // transaction set, so we know the exact closeTime against which to
+    // validate here -- 'nextCloseTime'.  (The _offset_, therefore, is
+    // the difference between 'nextCloseTime' and the last ledger close time.)
+    auto upperBoundCloseTimeOffset =
```

Comment:
As those behaviors are supposed to work in lockstep, it would be better to write the code like this:
```
    if (getHerderSCPDriver().curProtocolPreservesTxSetCloseTimeAffinity())
    {
        upperBoundCloseTimeOffset = nextCloseTime - lcl.header.scpValue.closeTime;
        lowerBoundCloseTimeOffset = upperBoundCloseTimeOffset;
    }
    else
    {
        upperBoundCloseTimeOffset = getUpperBoundCloseTimeOffset(mApp, lcl.header.scpValue.closeTime);
        lowerBoundCloseTimeOffset = 0;
    }
```

Reply:
Agreed and changed.